engine/src/main/java/org/camunda/bpm/engine/impl/pvm/runtime/PvmExecutionImpl.java

```diff
@@ -2263,4 +2267,26 @@ public IncidentHandler findIncidentHandler(String incidentType) {
     Map<String, IncidentHandler> incidentHandlers = Context.getProcessEngineConfiguration().getIncidentHandlers();
     return incidentHandlers.get(incidentType);
   }
+
+  public boolean isExecutingScopeLeafActivity() {
+    return isActive && getActivity() != null && getActivity().isScope() && activityInstanceId != null
+        && !(getActivity().getActivityBehavior() instanceof CompositeActivityBehavior);
+  }
+
+  /**
+   * This case is special, because the execution tree is different if an async after
+   * activity is left via transition (case 1) or not (case 2). In case 1, when the
+   * execution becomes async, the scope execution is already removed. In case 2 it is not.
+   *
+   * @return true if
+   * <ul>
+   *   <li>the execution is in asyncAfter state and completes
+   *   <li>leaves a scope activity
+   *   <li>completes the parent scope (i.e. does not leave via a transition
+   *   but propagates control to the parent)
+   * </ul>
+   */
+  public boolean isAsyncAfterScopeWithoutTransition() {
+    return activityInstanceId == null && activity.isScope() && !isActive;
```

Comment:
I tried to find any other circumstance where this turns `true` but shouldn't - I couldn't think of any. I really tried 😁 So to the best of my abilities, I think this should be fine.